src/card_operations/card_pair.c

```diff
@@ -0,0 +1,397 @@
+    LOG_CRITICAL("xxec %d:%d", ATECC_ERROR_BASE + status, __LINE__);
+    return ATECC_ERROR_BASE + status;
+  }
+
+  pair_data->data_len +=
+      ecdsa_sig_to_der(sig, pair_data->data + pair_data->data_len);
+
+  pair_data->is_paired = false;
+  return SUCCESS;
+}
+
+static uint32_t pair_card_postprocess(card_pairing_data_t *pair_data,
+                                      uint8_t card_number) {
+  HDNode guest_card_node;
+  SHA512_CTX ctx;
+  uint8_t keystore_index, digest[32] = {0}, buffer[64] = {0};
+  uint8_t public_key_uncompressed[65] = {0};
+  uint32_t index;
+  char xpub[112] = {'\0'};
+
+  base58_encode_check(get_card_root_xpub(),
+                      FS_KEYSTORE_XPUB_LEN,
+                      nist256p1_info.hasher_base58,
+                      xpub,
+                      112);
+  hdnode_deserialize_public(
+      (char *)xpub, 0x0488b21e, NIST256P1_NAME, &guest_card_node, NULL);
+
+  index = read_be(pair_data->data + 36);
+  hdnode_public_ckd(&guest_card_node, index);
+
+  index = read_be(pair_data->data + 40);
+  hdnode_public_ckd(&guest_card_node, index);
+
+  der_to_sig(pair_data->data + 44, buffer);
+  sha256_Raw(pair_data->data, 44, digest);
+  uint8_t status = ecdsa_verify_digest(
+      &nist256p1, guest_card_node.public_key, buffer, digest);
+  if (status) {
+    LOG_CRITICAL("xxec %d:%d", EXCEPTION_CARD_NOT_VERIFIED + status, __LINE__);
+    log_hex_array("resp", pair_data->data, 128);
+    log_hex_array("sig", buffer, sizeof(buffer));
+    mark_core_error_screen(ui_text_cannot_verify_card_contact_support);
+    return EXCEPTION_CARD_NOT_VERIFIED + status;
+  }
+
+  keystore_index = card_number - 1;
+  ecdsa_uncompress_pubkey(
+      &nist256p1, guest_card_node.public_key, public_key_uncompressed);
+  status = atecc_nfc_ecdh(&public_key_uncompressed[1], pair_data->data + 45);
+  if (ATCA_SUCCESS != status) {
+    LOG_CRITICAL("xxec %d:%d", ATECC_ERROR_BASE + status, __LINE__);
+    return ATECC_ERROR_BASE + status;
+  }
+
+  sha512_Init(&ctx);
+  sha512_Update(&ctx, pair_data->data + 45, 32);
+  sha512_Update(&ctx, pair_data->data + 4, 32);
+  sha512_Update(&ctx, pair_data->session_nonce, 32);
+  sha512_Final(&ctx, buffer);
+  set_keystore_pairing_key(
+      keystore_index, buffer, sizeof(buffer), FLASH_SAVE_LATER);
+  set_keystore_key_id(keystore_index, pair_data->data, 4, FLASH_SAVE_LATER);
+  set_keystore_used_status(keystore_index, 1, FLASH_SAVE_NOW);
+
+  return SUCCESS;
+}
+
+static uint32_t handle_pairing_success(card_operation_data_t *card_data,
+                                       card_pairing_data_t *pair_data) {
+  if (false == pair_data->is_paired) {
+    if (*(uint32_t *)get_family_id() == DEFAULT_UINT32_IN_FLASH) {
+      set_family_id_flash(card_data->nfc_data.family_id);
+    }
+
+    uint32_t error_status =
+        pair_card_postprocess(pair_data, card_data->nfc_data.tapped_card);
+    if (SUCCESS != error_status) {
+      return error_status;
+    }
+  }
+  return card_data->nfc_data.status;
```

Comment:
I am really concerned about clashing of error codes

Reply:
We've kept error masks defined in app_error.h, for this purpose.